packages/synapse-react-client/src/components/DataGrid/hooks/useGridUndo.tsx

```diff
@@ -0,0 +1,197 @@
+import { DataGridRow, Operation } from '../DataGridTypes'
+import { useCallback, useEffect, useState } from 'react'
+import { ModelChange } from '../utils/applyModelChange'
+import { Button, Menu, MenuItem } from '@mui/material'
+import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown'
+import { getInverseOperation } from '../utils/getInverseOperation'
+import { useStack } from './useStack'
+
+/**
+ * Represents a single undoable action in the grid
+ */
+export type UndoableAction = {
+  type: 'CREATE' | 'DELETE' | 'UPDATE'
+  rowIndex: number
+  previousValue?: DataGridRow
+  newValue?: DataGridRow
+}
+
+const MAX_UNDO_STEPS = 100
+/**
+ * Hook to manage undo functionality for a grid backed by a CRDT model.
+ *
+ * @param onApplyModelChange - Function to apply model changes and commit
+ */
+export function useGridUndo(onApplyModelChange: (change: ModelChange) => void) {
+  // Tracks the stack of undoable actions in LIFO order
+  const undoStack = useStack<UndoableAction>([], MAX_UNDO_STEPS)
+
+  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null)
+  const open = Boolean(anchorEl)
+
+  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
+    setAnchorEl(event.currentTarget)
+  }
+
+  const handleClose = () => setAnchorEl(null)
+
+  // Generate preview information for the undo menu to show users exactly what will be undone
+  // Returns: lastType (CREATE/UPDATE/DELETE), sameTypeCount (consecutive actions of same type),
+  // totalActions (all actions in stack).
+  const getUndoPreview = () => {
+    if (undoStack.isEmpty()) return null
+
+    const lastAction = undoStack.peek()
+    const lastType = lastAction.type
+
+    // Count consecutive actions of same type from the end
+    let sameTypeCount = 0
+    for (let i = undoStack.stack.length - 1; i >= 0; i--) {
+      if (undoStack.stack[i].type === lastType) {
+        sameTypeCount++
+      } else {
+        break
+      }
+    }
+
+    return {
+      lastType,
+      sameTypeCount,
+      totalActions: undoStack.size,
+    }
+  }
+
+  const undoPreview = getUndoPreview()
+
+  /**
+   * Undo one or more actions based on the chosen type:
+   * - 'lastAction': undo only the most recent action
+   */
+  const handleUndo = useCallback(
+    (handleClose?: () => void) => {
+      if (undoStack.isEmpty()) return
+
+      const actionToUndo = undoStack.pop()
+
+      if (actionToUndo) {
+        const inverseOperation = getInverseOperation(actionToUndo)
+
+        if (!inverseOperation) return
+
+        onApplyModelChange(inverseOperation)
+      }
+
+      handleClose?.()
+    },
+    [onApplyModelChange, undoStack],
```

Comment:
You can just reference the local `handleClose`
```suggestion
    () => {
      if (undoStack.isEmpty()) return

      const actionToUndo = undoStack.pop()

      if (actionToUndo) {
        const inverseOperation = getInverseOperation(actionToUndo)

        if (!inverseOperation) return

        onApplyModelChange(inverseOperation)
      }

      handleClose()
    },
    [onApplyModelChange, undoStack, handleClose],
```